File: tests/typesafety/python/deploy_mirror.py
# expect: passes
# title: The config file's grammar, copied into types that know nothing
from dataclasses import dataclass
from typing import cast

type Json = None | bool | int | float | str | list["Json"] | dict[str, "Json"]


@dataclass(frozen=True)
class Resource:
    kind: str
    name: str
    attributes: dict[str, Json]


def port_of(resource: Resource) -> int:
    return cast(int, resource.attributes["port"])


def replicas_of(resource: Resource) -> int:
    return cast(int, resource.attributes.get("replicas", 1))


def summary(resource: Resource) -> str:
    return f"{resource.name} on {port_of(resource)} x{replicas_of(resource)}"

File: tests/typesafety/typescript/deploy_mirror.ts
// expect: passes

type Json = null | boolean | number | string | Json[] | { [key: string]: Json };

type Resource = {
  readonly kind: string;
  readonly name: string;
  readonly attributes: { [key: string]: Json };
};

function portOf(resource: Resource): number {
  return resource.attributes.port as number;
}

function replicasOf(resource: Resource): number {
  return (resource.attributes.replicas ?? 1) as number;
}

export function summary(resource: Resource): string {
  return `${resource.name} on ${portOf(resource)} x${replicasOf(resource)}`;
}
